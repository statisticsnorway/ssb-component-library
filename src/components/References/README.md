References
========

> Wrapper for title, paragraph and link
### Usage


#### HTML

```html
<div class="ssb-references">
  <h5 class="ssb-title no-margin">Sources</h5>
  <a class="ssb-link" href=" " target="" rel="">
    <span class="link-text">Source link</span>
  </a>
</div>

<div class="ssb-references">
  <h5 class="ssb-title no-margin">Sources</h5>
  <span class="ssb-text-wrapper">Plain text</span>
  <a class="ssb-link" href=" " target="" rel="">
    <span class="link-text">Source link</span>
  </a>
</div>
```

#### React

```jsx harmony
const referenceList = [
  {
    href: ' ',
    label: 'Link 1',
  }, {
    href: ' ',
    label: 'Link 2',
  }, {
    href: ' ',
    label: 'Link 3',
    plainText: 'A paragraph',
  },
];

<References title="Sources" referenceList={referenceList} />
```

Available props:

| Name       | Type           | Description  |
| ---------- | ------------- | ----- |
| referenceList | arrayOf({ href, label, plainText }) | Required list of objects to render content |
| title | string| Header text |

__Dependencies__
 - [Link](../Link)
 - [Text](../Title)
 - [Title](../Title)
