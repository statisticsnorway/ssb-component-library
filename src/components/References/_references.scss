.ssb-references {
  display: inline-flex;
  flex-direction: column;
}
