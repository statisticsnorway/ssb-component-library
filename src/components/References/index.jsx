import React from 'react';
import PropTypes from 'prop-types';
import Link from '../Link';
import Text from '../Text';
import Title from '../Title';

const References = ({
	referenceList, title,
}) => (
	<div className="ssb-references">
		<Title size={5} className="no-margin">{title}</Title>
		{referenceList.map((ref, index) => (
			<div key={index}> {/* eslint-disable-line */}
				{ref.plainText && <Text className="no-margin">{ref.plainText}&nbsp;</Text>}
				<Link href={ref.href}>{ref.label}</Link><br />
			</div>
		))}
	</div>
);

References.defaultProps = {
	title: 'Kilder',
};

References.propTypes = {
	referenceList: PropTypes.arrayOf(
		PropTypes.shape({
			href: PropTypes.string.isRequired,
			label: PropTypes.string.isRequired,
			plainText: PropTypes.string,
		})
	),
	title: PropTypes.string,
};

export default References;
