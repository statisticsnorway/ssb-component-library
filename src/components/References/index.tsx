import React from 'react'
import Link from '../Link'
import Text from '../Text'

interface Item {
  href: string
  label: string
  plainText?: string
}
interface ReferencesProps {
  className?: string
  referenceList: Item[]
  title?: string
}

const References: React.FC<ReferencesProps> = ({ className, referenceList, title = 'Kilder' }) => (
  <div className={`ssb-references${className ? ` ${className}` : ''}`}>
    <span className='reference-header'>{title}</span>
    {referenceList.map((ref, index) => (
      // eslint-disable-next-line react/no-array-index-key
      <div key={index}>
        {ref.plainText && <Text className='no-margin'>{ref.plainText}&nbsp;</Text>}
        <Link href={ref.href}>{ref.label}</Link>
        <br />
      </div>
    ))}
  </div>
)

export default References
