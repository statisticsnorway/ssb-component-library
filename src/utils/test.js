import fs from 'fs';
import path from 'path';
import { render as rtlRender } from '@testing-library/react';

export * from '@testing-library/react';
export * from '@testing-library/jest-dom';

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8');

export function render(ui, options) {
	const view = rtlRender(ui, options);

	const styleElement = document.createElement('style');
	styleElement.innerHTML = styles;
	document.body.appendChild(styleElement);

	return view;
}
