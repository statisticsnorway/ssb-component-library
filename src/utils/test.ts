import fs from 'fs'
import path from 'path'
import { RenderOptions, RenderResult, render as rtlRender } from '@testing-library/react'
import { ReactNode } from 'react'

export * from '@testing-library/react'

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8')

export function render(ui: ReactNode, options?: RenderOptions): RenderResult {
  const view = rtlRender(ui, options)

  const styleElement = document.createElement('style')
  styleElement.innerHTML = styles
  document.body.appendChild(styleElement)

  return view
}
