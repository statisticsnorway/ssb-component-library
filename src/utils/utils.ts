// utils.ts
const classNames = (...classes: (string | boolean | undefined)[]): string => classes.filter(Boolean).join(' ')

export default classNames
