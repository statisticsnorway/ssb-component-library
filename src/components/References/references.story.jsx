import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import References from './index';

const referenceList = [
	{
		href: ' ',
		label: 'Link 1',
	}, {
		href: ' ',
		label: 'Link 2',
	}, {
		href: ' ',
		label: 'Link 3',
		plainText: 'A paragraph',
	},
];

const referenceList2 = [
	{
		href: ' ',
		label: 'Link 1',
	}, {
		href: ' ',
		label: 'Link 2',
	}, {
		href: ' ',
		label: 'Link 3',
	},
];

storiesOf('References', module).addDecorator(centered)
	.add('Regular', () => (
		<References title="Source" referenceList={referenceList2} />
	))
	.add('With paragraph', () => (
		<References title="Source" referenceList={referenceList} />
	));
