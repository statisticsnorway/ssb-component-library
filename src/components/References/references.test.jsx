import React from 'react'
import { render } from '../../utils/test'
import References from './index'

const referenceList = [
  {
    href: ' ',
    label: 'Link 1',
  },
  {
    href: ' ',
    label: 'Link 2',
  },
  {
    href: ' ',
    label: 'Link 3',
    plainText: 'A paragraph',
  },
]

describe('References component', () => {
  test('Matches the snapshot', () => {
    const { asFragment } = render(<References title='Source' referenceList={referenceList} />)
    expect(asFragment()).toMatchSnapshot()
  })
})
