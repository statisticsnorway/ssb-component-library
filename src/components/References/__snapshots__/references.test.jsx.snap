// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`References component Matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ssb-references"
  >
    <span
      class="reference-header"
    >
      Source
    </span>
    <div>
      <a
        class="ssb-link"
        href=" "
      >
        <span
          class="link-text"
        >
          Link 1
        </span>
      </a>
      <br />
    </div>
    <div>
      <a
        class="ssb-link"
        href=" "
      >
        <span
          class="link-text"
        >
          Link 2
        </span>
      </a>
      <br />
    </div>
    <div>
      <span
        class="ssb-text-wrapper no-margin"
      >
        A paragraphÂ 
      </span>
      <a
        class="ssb-link"
        href=" "
      >
        <span
          class="link-text"
        >
          Link 3
        </span>
      </a>
      <br />
    </div>
  </div>
</DocumentFragment>
`;
